File: plugin/skills/methodological-rules/scripts/update-rules.js
#!/usr/bin/env node
/**
 * update-rules.js
 * Called by PostPhaseTransition hook to update rule statuses
 *
 * Usage:
 *   node update-rules.js --project-path /path/to/project
 *
 * This script:
 * 1. Checks current phase
 * 2. Determines which rules need status updates
 * 3. Regenerates affected rules
 * 4. Logs changes to reflexivity journal
 */

const fs = require('fs');
const path = require('path');
const { execSync } = require('child_process');

function parseArgs() {
  const args = process.argv.slice(2);
  const parsed = {};
  for (let i = 0; i < args.length; i++) {
    if (args[i].startsWith('--')) {
      const key = args[i].replace('--', '');
      const value = args[i + 1] && !args[i + 1].startsWith('--') ? args[++i] : true;
      parsed[key] = value;
    }
  }
  return parsed;
}

function logToJournal(projectPath, message) {
  const journalPath = path.join(projectPath, '.interpretive-orchestration', 'reflexivity-journal.md');
  if (!fs.existsSync(journalPath)) return;

  const entry = `
---

### Phase Transition - Rules Updated
**Date:** ${new Date().toISOString().split('T')[0]}
**Time:** ${new Date().toTimeString().split(' ')[0]}

${message}

---
`;

  try {
    fs.appendFileSync(journalPath, entry);
  } catch (e) {
    // Non-critical
  }
}

// Main execution
const args = parseArgs();

if (!args['project-path']) {
  console.error(JSON.stringify({
    success: false,
    error: 'Missing required argument: --project-path'
  }));
  process.exit(1);
}

const projectPath = path.resolve(args['project-path']);
const rulesDir = path.join(projectPath, '.claude', 'rules');

// First, check current phase status
const checkPhaseScript = path.join(__dirname, 'check-phase.js');
let phaseInfo;

try {
  const output = execSync(`node "${checkPhaseScript}" --project-path "${projectPath}"`, {
    encoding: 'utf8'
  });
  phaseInfo = JSON.parse(output);
} catch (e) {
  console.error(JSON.stringify({
    success: false,
    error: `Failed to check phase: ${e.message}`
  }));
  process.exit(1);
}

// Check if any rules exist that need updating
const existingRules = [];
if (fs.existsSync(rulesDir)) {
  const files = fs.readdirSync(rulesDir);
  for (const file of files) {
    if (file.endsWith('.md')) {
      existingRules.push(file.replace('.md', ''));
    }
  }
}

// Determine what changed
const relaxedRules = phaseInfo.rules_should_relax.map(r => r.name);
const rulesNeedingUpdate = existingRules.filter(ruleName => {
  // Read current rule to check its status
  const rulePath = path.join(rulesDir, `${ruleName}.md`);
  try {
    const content = fs.readFileSync(rulePath, 'utf8');
    const isCurrentlyActive = content.includes('**Rule Status:** ACTIVE');
    const shouldBeRelaxed = relaxedRules.includes(ruleName);

    // Need update if status mismatch
    return isCurrentlyActive && shouldBeRelaxed;
  } catch (e) {
    return false;
  }
});

if (rulesNeedingUpdate.length === 0) {
  console.log(JSON.stringify({
    success: true,
    message: 'No rules need updating',
    current_phase: phaseInfo.current_phase,
    rules_checked: existingRules.length
  }));
  process.exit(0);
}

// Regenerate all rules (simplest approach - ensures consistency)
const generateScript = path.join(__dirname, 'generate-rules.js');
let generateResult;

try {
  const output = execSync(`node "${generateScript}" --project-path "${projectPath}"`, {
    encoding: 'utf8'
  });
  generateResult = JSON.parse(output);
} catch (e) {
  console.error(JSON.stringify({
    success: false,
    error: `Failed to regenerate rules: ${e.message}`
  }));
  process.exit(1);
}

// Log the transition
const transitionMessage = `**Phase Transition Detected**

**New Phase:** ${phaseInfo.current_phase}

**Rules Relaxed:**
${rulesNeedingUpdate.map(r => `- ${r}`).join('\n') || '- None'}

**Rules Still Active:**
${phaseInfo.rules_still_active.map(r => `- ${r.name} (until ${r.relaxes_at})`).join('\n') || '- None'}

This is a natural part of your analytical progression. Rules relax when methodologically appropriate.`;

logToJournal(projectPath, transitionMessage);

// Output result
console.log(JSON.stringify({
  success: true,
  current_phase: phaseInfo.current_phase,
  rules_updated: rulesNeedingUpdate,
  rules_regenerated: generateResult.rules_generated?.length || 0,
  message: `Phase transition: ${rulesNeedingUpdate.length} rule(s) relaxed`
}, null, 2));

process.exit(0);

File: plugin-kimi/.agents/skills/qual-methodological-rules/scripts/update_rules.py
#!/usr/bin/env python3
"""
update_rules.py
Update rule statuses after phase transitions

Usage:
    python3 update_rules.py --project-path /path/to/project

This script:
1. Checks current phase
2. Determines which rules need status updates
3. Regenerates affected rules
4. Logs changes to reflexivity journal
"""

import argparse
import json
import sys
from pathlib import Path
from datetime import datetime

# Import sibling scripts
try:
    from check_phase import check_phase
    from generate_rules import generate_rules
except ImportError:
    # Fallback: add current directory to path
    sys.path.insert(0, str(Path(__file__).parent))
    from check_phase import check_phase
    from generate_rules import generate_rules

# Import qual-shared
sys.path.insert(0, str(Path(__file__).parent.parent.parent / "qual-shared" / "scripts"))
try:
    from conversation_logger import ConversationLogger
except ImportError:
    ConversationLogger = None


def log_to_journal(project_path: Path, message: str):
    """
    Log phase transition to reflexivity journal

    Args:
        project_path: Path to project
        message: Message to log
    """
    # Use ConversationLogger if available
    if ConversationLogger:
        try:
            logger = ConversationLogger(str(project_path))
            logger.log_event("phase_transition_rules_updated", {"message": message})
            return
        except:
            pass  # Fall through to manual append

    # Fallback: Manual journal append
    journal_path = project_path / ".interpretive-orchestration" / "reflexivity-journal.md"
    if not journal_path.exists():
        return

    today = datetime.now().strftime('%Y-%m-%d')
    time = datetime.now().strftime('%H:%M:%S')

    entry = f"""
---

### Phase Transition - Rules Updated
**Date:** {today}
**Time:** {time}

{message}

---
"""

    try:
        with open(journal_path, 'a') as f:
            f.write(entry)
    except Exception:
        # Non-critical
        pass


def update_rules(project_path: str):
    """
    Update rule statuses after phase transition

    Args:
        project_path: Path to project directory

    Returns:
        dict: Update result with status information
    """
    project_path = Path(project_path).resolve()

    # Check current phase
    phase_info = check_phase(str(project_path))
    if not phase_info.get('success'):
        return phase_info

    # Load existing rules to check what needs updating
    rules_file = project_path / ".interpretive-orchestration" / "methodological-rules.json"
    existing_rules = []

    if rules_file.exists():
        try:
            with open(rules_file, 'r') as f:
                existing_data = json.load(f)
                existing_rules = existing_data.get('rules', [])
        except Exception:
            existing_rules = []

    # Determine what changed
    relaxed_rule_names = [r['name'] for r in phase_info['rules_should_relax']]
    rules_needing_update = []

    for rule in existing_rules:
        is_currently_active = rule.get('status') == 'active'
        should_be_relaxed = rule.get('rule_id') in relaxed_rule_names

        # Need update if status mismatch
        if is_currently_active and should_be_relaxed:
            rules_needing_update.append(rule.get('rule_id', 'unknown'))

    if not rules_needing_update:
        return {
            "success": True,
            "message": "No rules need updating",
            "current_phase": phase_info['current_phase'],
            "rules_checked": len(existing_rules)
        }

    # Regenerate all rules (ensures consistency)
    generate_result = generate_rules(str(project_path))

    if not generate_result.get('success'):
        return {
            "success": False,
            "error": f"Failed to regenerate rules: {generate_result.get('error', 'Unknown error')}"
        }

    # Log the transition
    active_rules_text = '\n'.join(
        f"- {r['name']} (until {r['relaxes_at']})"
        for r in phase_info['rules_still_active']
    ) if phase_info['rules_still_active'] else '- None'

    relaxed_rules_text = '\n'.join(
        f"- {rule_id}"
        for rule_id in rules_needing_update
    ) if rules_needing_update else '- None'

    transition_message = f"""**Phase Transition Detected**

**New Phase:** {phase_info['current_phase']}

**Rules Relaxed:**
{relaxed_rules_text}

**Rules Still Active:**
{active_rules_text}

This is a natural part of your analytical progression. Rules relax when methodologically appropriate."""

    log_to_journal(project_path, transition_message)

    # Output result
    return {
        "success": True,
        "current_phase": phase_info['current_phase'],
        "rules_updated": rules_needing_update,
        "rules_regenerated": len(generate_result.get('rules', [])),
        "message": f"Phase transition: {len(rules_needing_update)} rule(s) relaxed"
    }


def main():
    """CLI entry point"""
    parser = argparse.ArgumentParser(
        description="Update rule statuses after phase transition"
    )
    parser.add_argument(
        '--project-path',
        type=str,
        required=True,
        help='Path to project directory'
    )

    args = parser.parse_args()

    result = update_rules(args.project_path)

    print(json.dumps(result, indent=2))

    return 0 if result.get('success') else 1


if __name__ == '__main__':
    sys.exit(main())
